### PR TITLE
Test HAWK graphical user interface using selenium

### PR DESCRIPTION
**Pull request by alvarocarvajald**
This PR adds a test for the HAWK GUI using a python test script with selenium, [geckodriver 0.24](https://github.com/mozilla/geckodriver/releases) and firefox. It includes the `ha/hawk_gui` module to schedule the test, the python script itself, as well as the necessary modifications in HA barriers and HA test schedule to use the test.

New variable `HAWKGUI_TEST_ROLE` is introduced. Set it to `client` in a job with access to a graphical console in conjunction with `CLUSTER_NAME` and `HA_CLUSTER` to schedule the client side test which will connect to the node 1 of the cluster with firefox and perform the selenium test. The cluster nodes tests should also include `HAWKGUI_TEST_ROLE` but set to something different than `client` in order for the `HAWK_GUI_*` barriers to be correctly setup.

All tests/nodes should be configured in multi-machine mode with their corresponding support server and network configuration.

- Related ticket: https://progress.opensuse.org/issues/30270
- Needles: https://gitlab.suse.de/openqa/os-autoinst-needles-sles/merge_requests/1102
- Verification runs:

1. Single SLED test with external SLES+HA 15-GM cluster: http://mango.suse.de/tests/824
2. MM test node1 in SLES+HA 15-SP1: http://1b147.qa.suse.de/tests/5241
3. MM test node2 in SLES+HA 15-SP1: http://1b147.qa.suse.de/tests/5242
4. MM test HAWK test node in SLES 15-SP1: http://1b147.qa.suse.de/tests/5243
5. MM test support server: http://1b147.qa.suse.de/tests/5240

### CODE REVIEW

File: data/ha/hawk_test.py [hunk 1]
```diff
@@ -0,0 +1,829 @@
...
+def is_ssh(s):
+    clase = str(type(s)).split(' ')[1][1:26]
+    if clase != 'paramiko.client.SSHClient':
+        print('ERROR: SSH object must be of type paramiko.client.SSHClient. Got: [%s]' % type(s))
+        return False
+    return True
+
+### Logging and results
+
+def logresults(filename):
```
Comment (asmorodskyi) [inline data/ha/hawk_test.py]: why not to use something native ? somrthing like https://docs.python.org/3/library/logging.html ?

File: data/ha/hawk_test.py [hunk 2]
```diff
@@ -0,0 +1,829 @@
...
+        # Default to 1 if no int argument
+        return 1
+
+def checkstr(s):
+    if isinstance(s, str):
+        return s
+    else:
+        return str(s)
+
+def is_driver(d):
```
Comment (asmorodskyi) [inline data/ha/hawk_test.py]: I think this function just creating boilerplate code which does not bring any value, and in some cases would create danger of swallow errors , for example only brief look point me to line 242 where you will get always passed test case when you pass not webdriver instance. No need to be so defensive exceptions are good they point you to root cause of the problem

**Comment by coolo** [comment]
Just curious: Is this python script a copy or developed for this?

File: data/ha/hawk_test.py [hunk 3]
```diff
@@ -0,0 +1,829 @@
...
+TARGET_ROLE_FORMAT = ('//select[contains(@class, "form-control") and '
+                      'contains(@name, "%s[meta][target-role]")]')
+STONITH_MAINT_OFF = ('//a[contains(@href, "stonith-sbd") and '
+                     'contains(@title, "Disable Maintenance Mode")]')
+
+# Functions
+
+### Functions to check types and classes
+
+def checkint(i):
```
Comment (asmorodskyi) [inline data/ha/hawk_test.py]: the same would be with checkint and checkstr functions I would prefer to get exception instead hiding the fact that I am passing "pinapple" to function which expect some integer

File: data/ha/hawk_test.py [hunk 4]
```diff
@@ -0,0 +1,829 @@
...
+                    help='Prefix to add to Resources created during the test')
+parser.add_argument('-t', '--test-version', type=str, default='',
+                    help='Test version. Ex: 12-SP3, 12-SP4, 15, 15-SP1')
+parser.add_argument('-s', '--secret', type=str, default='',
+                    help='root SSH Password of the HAWK node')
+parser.add_argument('-r', '--results', type=str, default='',
+                    help='Generate hawk_test.results file')
+args = parser.parse_args()
+if not args.browser:
+    print('ERROR: must specify a browser with --browser')
```
Comment (asmorodskyi) [inline data/ha/hawk_test.py]: ArgumentParser has special property of argument something like required= true . this will allow you to drop this if at all and let ArgumentParser do the job

File: data/ha/hawk_test.py [hunk 5]
```diff
@@ -0,0 +1,829 @@
...
+    print("TEST: Will click on Constraints, Nodes, Tags, Alerts and Fencing")
+    check_and_click_by_xpath(browser, timeout_scale, 17,
+                             "while checking around edit configuration",
+                             [HREF_CONSTRAINTS, HREF_NODES, HREF_TAGS, HREF_ALERTS, HREF_FENCING])
+    if RETURN_VALUE != 17:
+        set_test_status('click_around_edit_conf', 'passed')
+
+# Set STONITH/sbd in maintenance. Assumes stonith-sbd resource is the last one listed on the
+# resources table
+def set_stonith_maintenance(browser, timeout_scale):
```
Comment (asmorodskyi) [inline data/ha/hawk_test.py]: would be nice to come up with some naming pattern , like all test cases end up or starting from some word

File: data/ha/hawk_test.py [hunk 6]
```diff
@@ -0,0 +1,829 @@
...
+    global RETURN_VALUE
+    timeout_scale = checkint(timeout_scale)
+    # wait for page to fully load
+    time.sleep(timeout_scale)
```
Comment (asmorodskyi) [inline data/ha/hawk_test.py]: you are sleeping before calling this function and you are sleeping inside , also problem of waiting page to load solved already within a webdriver you don't need to waste time on this https://stackoverflow.com/questions/36590274/selenium-how-to-wait-until-page-is-completely-loaded

File: data/ha/hawk_test.py [hunk 7]
```diff
@@ -0,0 +1,829 @@
...
+myprimitive = args.prefix.lower() + 'cool_primitive'
+myclone = args.prefix.lower() + 'cool_clone'
+mygroup = args.prefix.lower() + 'cool_group'
+
+# Variables for texts that change between chrome and firefox drivers
+troubleshoot = link_by_browser(browser, 'troubleshooting')
+monitoring = link_by_browser(browser, 'monitoring')
+
+# Tests to perform
+browser.get(mainlink)
```
Comment (asmorodskyi) [inline data/ha/hawk_test.py]: below is some dirty mix of tests functions called and I suspect depend on each other and even some action happening in between. 
Each test case should be started with new instance of web driver which will give you clean cache and no side-effects of other tests . Otherwise you will have hard times to maintain all this , also you will have hard times to be confident that if test C failed you really have problem  with feature C and not that test A leave some crap behind which cause feature C not to work .

**Comment by alvarocarvajald** [comment]
> Just curious: Is this python script a copy or developed for this?

This started as 2 python+selenium scripts provided by QAM, one to test hawk with chrome/chromium using chromedriver, and the other doing the same but with firefox and geckodriver 0.17. Both scripts were merged into one. Additionally, we also got from QAM a [golang+agouti script](https://github.com/atighineanu/hawktest_ha_agouti) as an alternative to the python+selenium scripts, but the test coverage of both sets of scripts (python vs. golang) was different, so I decided to merge that as well into the merged python script.

So, long story short, I would say this was developed for this.

Should this be here? One of the reasons why I included it in the PR was to get some peer reviews from outside the team (thanks @asmorodskyi), but I am not sure about adding close to 800 lines in python to this repository either.

**Comment by coolo** [comment]
> Should this be here? One of the reasons why I included it in the PR was to get some peer reviews from outside the team (thanks @asmorodskyi), but I am not sure about adding close to 800 lines in python to this repository either.

It depends. If it's maintained elsewhere: no, don't copy it here but find other means to install it into the SUT.
If it's maintained here: feel free to do so.

File: data/ha/hawk_test.py (hunk 1 above)
Comment (alvarocarvajald) [reply data/ha/hawk_test.py]: Looks good. Don't think it's a good alternative for this function (as a `json.dumps()` of the `dict` is simpler), but could help elsewhere in the script. Will look into. (should've named this function `saveresults` instead of `logresults` to avoid confusion)

File: data/ha/hawk_test.py (hunk 2 above)
Comment (alvarocarvajald) [reply data/ha/hawk_test.py]: The code after line 242 (L257) is indeed wrong. Fixing it.

File: data/ha/hawk_test.py (hunk 3 above)
Comment (alvarocarvajald) [reply data/ha/hawk_test.py]: That's precisely the thing: I am trying to avoid exceptions whenever possible. When there's an exception, the script aborts completely; if this happens after the browser has already been opened, the browser remains open without a controlling driver because it does not reach the call to `browser.quit()`. I guess I could add a call to quit the browser when `except Exception:`, but figured it was not a good practice.

File: data/ha/hawk_test.py (hunk 4 above)
Comment (alvarocarvajald) [reply data/ha/hawk_test.py]: Fixing it.

File: data/ha/hawk_test.py (hunk 7 above)
Comment (alvarocarvajald) [reply data/ha/hawk_test.py]: Wouldn't this mean that a new session (including logging in to hawk) is started with each new instance of webdriver? It is a good idea, and I figure it will help with some tests particularly when using this same script to test hawk versions in 12-SP*, but I worry about how much more time this would add to the test. Currently it's taking close to 7 minutes (which is why I'm giving the `ha/hawk_gui` test module 15 minutes to work). Will look into it though, as it'll be easier to maintain. Thanks!

File: data/ha/hawk_test.py (hunk 6 above)
Comment (alvarocarvajald) [reply data/ha/hawk_test.py]: Fixed duplicate calls to `sleep()`. Will look into https://stackoverflow.com/questions/36590274/selenium-how-to-wait-until-page-is-completely-loaded

Comment (alvarocarvajald) [reply data/ha/hawk_test.py]: Regarding waiting for the page to load, [this](https://stackoverflow.com/questions/26566799/wait-until-page-is-loaded-with-selenium-webdriver-for-python) seems to be the python way, and I am already using that function within `find_element`, so this change should be straightforward. Will push the changes shortly.

File: data/ha/hawk_test.py (hunk 7 above)
Comment (asmorodskyi) [reply data/ha/hawk_test.py]: every solution has it's pros and cons. Having new instance of webdriver would slow you down. But I don't think that it will slow you down dramatically. Also you will get huge benefits - additionally to what I already mention you can feel more relaxed about failures in the test. Each your test would  be capable to fail only itself but not whole test suite

**Comment by alvarocarvajald** [comment]
Just pushed a commit applying the changes requested by @asmorodskyi. Still need to look into using a different instance of webdriver for each test, and into python's logging facility, but since those could lead to major changes in the code will probably do it in a separate PR.

Updated verification runs:

1. Single machine, external cluster: http://mango.suse.de/tests/826
2. MM, node1: http://1b147.qa.suse.de/tests/5245
3. MM, node2: http://1b147.qa.suse.de/tests/5246
4. MM, hawk test: http://1b147.qa.suse.de/tests/5247
5. MM, support server: http://1b147.qa.suse.de/tests/5244

**Comment by asmorodskyi** [comment]
So it is your game and looks like you ready to move on no matter that python code is far from ideal. I will merge it ( I not doomed to support it :)  )  But please consider following changes in nearest future otherwise this file will bring you a lot of pain : 
1. 1000-lines files is really bad thing . This file should be split into several . ideally 1 file for helper functions and one file per test suite 
2. drop is_driver/checkstr/checkint functions they are really evil. Don't know where you have read that catching browser exceptions is bad thing but it is not. What reasoning they gave BTW ? Can you point me to some article which explain this ? Wrap whole test into try/catch and unconditionally exit is much better than this evil_functions 
3. Isolate each test case logic . Make every test independent from others 

Good luck :)

**Comment by alvarocarvajald** [comment]
Thank you very much for the feedback.

> 1. 1000-lines files is really bad thing . This file should be split into several . ideally 1 file for helper functions and one file per test suite

Will do.

> 2. drop is_driver/checkstr/checkint functions they are really evil. Don't know where you have read that catching browser exceptions is bad thing but it is not. What reasoning they gave BTW ? Can you point me to some article which explain this ? Wrap whole test into try/catch and unconditionally exit is much better than this evil_functions

Not that I read or think that catching exceptions is a bad thing, but more that I did not want the script to abort in the background while the focus is on the firefox window in the foreground, and that I did not want to have a catch all exceptions (`except Exception`) in the code. Making each test independent and catching specific exceptions for each would be the way to go.

> 3. Isolate each test case logic . Make every test independent from others

Will do.

Again, thank you for the help.

**Comment by juadk** [comment]
I've never used the selenium module but sounds good to me, cool new test.
